File: tests/test_lexer.py
"""Tests for the PML lexer."""

import pytest

from pml.errors import PMLSyntaxError
from pml.lexer import Lexer, TokenType


class TestBasicTokens:
    def test_integers(self):
        tokens = Lexer("42 -10 +7").tokenize()
        assert tokens[0].type == TokenType.INTEGER
        assert tokens[0].value == "42"
        assert tokens[1].type == TokenType.INTEGER
        assert tokens[1].value == "-10"
        assert tokens[2].type == TokenType.INTEGER
        assert tokens[2].value == "+7"

    def test_floats(self):
        tokens = Lexer("3.14 -0.5").tokenize()
        assert tokens[0].type == TokenType.FLOAT
        assert tokens[0].value == "3.14"
        assert tokens[1].type == TokenType.FLOAT
        assert tokens[1].value == "-0.5"

    def test_strings(self):
        tokens = Lexer('"hello" "world"').tokenize()
        assert tokens[0].type == TokenType.STRING
        assert tokens[0].value == "hello"
        assert tokens[1].type == TokenType.STRING
        assert tokens[1].value == "world"

    def test_string_escapes(self):
        tokens = Lexer(r'"a\"b" "c\\d"').tokenize()
        assert tokens[0].value == 'a"b'
        assert tokens[1].value == "c\\d"

    def test_symbols(self):
        tokens = Lexer("circle my-var + ->").tokenize()
        assert tokens[0].type == TokenType.SYMBOL
        assert tokens[0].value == "circle"
        assert tokens[1].value == "my-var"
        assert tokens[2].value == "+"
        assert tokens[3].value == "->"

    def test_booleans(self):
        tokens = Lexer("#t #f").tokenize()
        assert tokens[0].type == TokenType.BOOLEAN
        assert tokens[0].value == "#t"
        assert tokens[1].type == TokenType.BOOLEAN
        assert tokens[1].value == "#f"

    def test_keywords(self):
        tokens = Lexer(":fill :stroke-width").tokenize()
        assert tokens[0].type == TokenType.KEYWORD
        assert tokens[0].value == "fill"
        assert tokens[1].type == TokenType.KEYWORD
        assert tokens[1].value == "stroke-width"

    def test_parens(self):
        tokens = Lexer("(+ 1 2)").tokenize()
        assert tokens[0].type == TokenType.LPAREN
        assert tokens[-2].type == TokenType.RPAREN
        assert tokens[-1].type == TokenType.EOF


class TestComments:
    def test_line_comment(self):
        tokens = Lexer("42 ; this is a comment\n10").tokenize()
        assert len(tokens) == 3  # 42, 10, EOF
        assert tokens[0].value == "42"
        assert tokens[1].value == "10"

    def test_only_comment(self):
        tokens = Lexer("; just a comment").tokenize()
        assert len(tokens) == 1
        assert tokens[0].type == TokenType.EOF


class TestSyntaxSugar:
    def test_quote(self):
        tokens = Lexer("'hello").tokenize()
        assert tokens[0].type == TokenType.QUOTE

    def test_quasiquote(self):
        tokens = Lexer("`hello").tokenize()
        assert tokens[0].type == TokenType.QUASIQUOTE

    def test_unquote(self):
        tokens = Lexer(",x").tokenize()
        assert tokens[0].type == TokenType.UNQUOTE

    def test_unquote_splicing(self):
        tokens = Lexer(",@xs").tokenize()
        assert tokens[0].type == TokenType.UNQUOTE_SPLICE


class TestLineTracking:
    def test_multiline(self):
        source = "(define x 1)\n(define y 2)"
        tokens = Lexer(source).tokenize()
        # First token on line 1
        assert tokens[0].line == 1
        # "define" on line 2
        define_token = [t for t in tokens if t.value == "define"][1]
        assert define_token.line == 2


class TestErrors:
    def test_unterminated_string(self):
        with pytest.raises(PMLSyntaxError, match="Unterminated string"):
            Lexer('"hello').tokenize()

File: tests/test_lexer.cpp
// ==========================================================================================================================================================================================================================================═
// PML Lexer Tests — Google Test suite for pml::Lexer
// ==========================================================================================================================================================================================================================================═

#include <gtest/gtest.h>
#include "pml/frontend/lexer.h"

#include <string>
#include <vector>

namespace pml {
namespace {

// Helper: tokenize source and return the token vector (asserts success).
std::vector<Token> tokenize_ok(const std::string& source) {
    auto result = Lexer(source, "<test>").tokenize();
    EXPECT_TRUE(result.has_value())
        << "tokenize() failed: " << result.error().message;
    return std::move(*result);
}

// ----─ Integer tokens --------------------------------------------------------------------------------------------------------------------

TEST(Lexer, IntegerLiteral) {
    auto tokens = tokenize_ok("42");
    ASSERT_EQ(tokens.size(), 2u);  // INTEGER + EOF
    EXPECT_EQ(tokens[0].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[0].value, "42");
}

TEST(Lexer, NegativeInteger) {
    auto tokens = tokenize_ok("-5");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[0].value, "-5");
}

// ----─ Float tokens ------------------------------------------------------------------------------------------------------------------------

TEST(Lexer, FloatLiteral) {
    auto tokens = tokenize_ok("3.14");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::FLOAT);
    EXPECT_EQ(tokens[0].value, "3.14");
}

// ----─ String tokens --------------------------------------------------------------------------------------------------------------------─

TEST(Lexer, StringLiteral) {
    auto tokens = tokenize_ok("\"hello\"");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::STRING);
    EXPECT_EQ(tokens[0].value, "hello");
}

TEST(Lexer, StringEscapeNewline) {
    auto tokens = tokenize_ok("\"hello\\nworld\"");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::STRING);
    EXPECT_EQ(tokens[0].value, "hello\nworld");
}

// ----─ Symbol tokens --------------------------------------------------------------------------------------------------------------------─

TEST(Lexer, SymbolWithDash) {
    auto tokens = tokenize_ok("foo-bar");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[0].value, "foo-bar");
}

// ----─ Boolean tokens --------------------------------------------------------------------------------------------------------------------

TEST(Lexer, BooleanTrue) {
    auto tokens = tokenize_ok("#t");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::BOOLEAN);
    EXPECT_EQ(tokens[0].value, "#t");
}

TEST(Lexer, BooleanFalse) {
    auto tokens = tokenize_ok("#f");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::BOOLEAN);
    EXPECT_EQ(tokens[0].value, "#f");
}

// ----─ Keyword tokens --------------------------------------------------------------------------------------------------------------------

TEST(Lexer, Keyword) {
    auto tokens = tokenize_ok(":key");
    ASSERT_EQ(tokens.size(), 2u);
    EXPECT_EQ(tokens[0].type, TokenType::KEYWORD);
    EXPECT_EQ(tokens[0].value, "key");  // stored without leading ':'
}

// ----─ Parenthesized expression ------------------------------------------------------------------------------------------------

TEST(Lexer, ParenthesizedExpression) {
    auto tokens = tokenize_ok("(+ 1 2)");
    // Expected: LPAREN, SYMBOL("+"), INTEGER("1"), INTEGER("2"), RPAREN, EOF
    ASSERT_EQ(tokens.size(), 6u);
    EXPECT_EQ(tokens[0].type, TokenType::LPAREN);
    EXPECT_EQ(tokens[1].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[1].value, "+");
    EXPECT_EQ(tokens[2].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[2].value, "1");
    EXPECT_EQ(tokens[3].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[3].value, "2");
    EXPECT_EQ(tokens[4].type, TokenType::RPAREN);
    EXPECT_EQ(tokens[5].type, TokenType::END_OF_FILE);
}

// ----─ Comments --------------------------------------------------------------------------------------------------------------------------------

TEST(Lexer, LineComment) {
    auto tokens = tokenize_ok("; comment\n42");
    ASSERT_EQ(tokens.size(), 2u);  // INTEGER("42") + EOF
    EXPECT_EQ(tokens[0].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[0].value, "42");
}

// ----─ Quote sugar ------------------------------------------------------------------------------------------------------------------------─

TEST(Lexer, QuoteSugar) {
    auto tokens = tokenize_ok("'x");
    ASSERT_EQ(tokens.size(), 3u);  // QUOTE, SYMBOL, EOF
    EXPECT_EQ(tokens[0].type, TokenType::QUOTE);
    EXPECT_EQ(tokens[1].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[1].value, "x");
}

// ----─ Quasiquote / unquote ----------------------------------------------------------------------------------------------------─

TEST(Lexer, QuasiquoteAndUnquote) {
    auto tokens = tokenize_ok("`(1 ,x)");
    // Expected: QUASIQUOTE, LPAREN, INTEGER("1"), UNQUOTE, SYMBOL("x"), RPAREN, EOF
    ASSERT_EQ(tokens.size(), 7u);
    EXPECT_EQ(tokens[0].type, TokenType::QUASIQUOTE);
    EXPECT_EQ(tokens[1].type, TokenType::LPAREN);
    EXPECT_EQ(tokens[2].type, TokenType::INTEGER);
    EXPECT_EQ(tokens[2].value, "1");
    EXPECT_EQ(tokens[3].type, TokenType::UNQUOTE);
    EXPECT_EQ(tokens[4].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[4].value, "x");
    EXPECT_EQ(tokens[5].type, TokenType::RPAREN);
    EXPECT_EQ(tokens[6].type, TokenType::END_OF_FILE);
}

// ----─ Empty input ------------------------------------------------------------------------------------------------------------------------─

TEST(Lexer, EmptyInput) {
    auto tokens = tokenize_ok("");
    ASSERT_EQ(tokens.size(), 1u);
    EXPECT_EQ(tokens[0].type, TokenType::END_OF_FILE);
}

// ----─ Unterminated string --------------------------------------------------------------------------------------------------------─

TEST(Lexer, UnterminatedString) {
    auto result = Lexer("\"hello", "<test>").tokenize();
    ASSERT_FALSE(result.has_value());
    EXPECT_EQ(result.error().code, ErrorCode::PMLSyntaxError);
}

// ----─ Line tracking --------------------------------------------------------------------------------------------------------------------─

TEST(Lexer, LineTracking) {
    // "(a\nb)" — '(' on line 1, 'a' on line 1, 'b' on line 2, ')' on line 2
    auto tokens = tokenize_ok("(a\nb)");
    ASSERT_GE(tokens.size(), 5u);  // LPAREN, SYMBOL, SYMBOL, RPAREN, EOF

    EXPECT_EQ(tokens[0].type, TokenType::LPAREN);
    EXPECT_EQ(tokens[0].line, 1);

    EXPECT_EQ(tokens[1].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[1].value, "a");
    EXPECT_EQ(tokens[1].line, 1);

    EXPECT_EQ(tokens[2].type, TokenType::SYMBOL);
    EXPECT_EQ(tokens[2].value, "b");
    EXPECT_EQ(tokens[2].line, 2);

    EXPECT_EQ(tokens[3].type, TokenType::RPAREN);
    EXPECT_EQ(tokens[3].line, 2);
}

}  // namespace
}  // namespace pml
